modules/options/base.go

```diff
@@ -0,0 +1,34 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package options
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+)
+
+func walkAssetDir(root string, callback func(path, name string, d fs.DirEntry, err error) error) error {
+	if err := filepath.WalkDir(root, func(path string, d fs.DirEntry, err error) error {
+		name := path[len(root):]
+		if len(name) > 0 && name[0] == '/' {
+			name = name[1:]
+		}
+		if err != nil {
```

Comment:
We likely want to check the `err` as one of the first things. Not sure what we can assume when a error occur(is path empty or not)

Reply:
The `path` component the `WalkDirFunc` is ALWAYS set - the docs do say this but we can look at the source code to examine this

```go
func WalkDir(root string, fn fs.WalkDirFunc) error {
	info, err := os.Lstat(root)
	if err != nil {
		err = fn(root, nil, err)
	} else {
		err = walkDir(root, &statDirEntry{info}, fn) 
	}
	if err == SkipDir {
		return nil
	}
	return err
}
```

Only one line here has `err != nil`: 

```go
		err = fn(root, nil, err) 
```

but you can see that this implies that the `path == root` and therefore this code is correct.

Then we look at `walkDir`:

```go
// walkDir recursively descends path, calling walkDirFn.
func walkDir(path string, d fs.DirEntry, walkDirFn fs.WalkDirFunc) error {
	if err := walkDirFn(path, d, nil); err != nil || !d.IsDir() { 
		if err == SkipDir && d.IsDir() {
			// Successfully skipped directory.
			err = nil
		}
		return err
	}

	dirs, err := readDir(path)
	if err != nil {
		// Second call, to report ReadDir error.
		err = walkDirFn(path, d, err)
		if err != nil {
			return err
		}
	}

	for _, d1 := range dirs {
		path1 := Join(path, d1.Name())
		if err := walkDir(path1, d1, walkDirFn); err != nil {
			if err == SkipDir {
				break
			}
			return err
		}
	}
	return nil
}
```

Again only one line has `err != nil`:

```go
		err = walkDirFn(path, d, err)
```

and again the `path` is always set (and will always have root as a prefix)

We calculate the name before checking the error because the error code uses this name in its logging.

It is therefore safe to defer checking the error until after we have done this calculation.